Location upload refactor

esoergel opened the pull request:
## Product-facing changes

* `location_id` is now a required field in order to update a location.  Previously, users had issues where they'd intend to add a new location, but the site_code they specified was already taken.  The old behavior was to update the existing location that had that site_code, whereas now it'll fail.  This ID requirement is consistent with user and case uploads.
* Location restricted users can now access the location uploader.  Restricted users can include any location they like in the upload, but they may only modify accessible ones.
* Restricted users may not modify location types.
* Restricted and non-restricted users have the option to download an excel file with just headers in it, to use as a template for adding new locations.  Restricted users currently see _only_ this file, so they basically can't edit existing locations in bulk.  The use case is primarily for bulk addition of new locations.

## Notes for reviewers
This is a big PR, I'd recommend going commit-by-commit.  There's one big commits, but most should be pretty digestible.  Let me know if I can clarify anything.  I'm treating this PR as a long-running feature branch, and PRing additions into this branch, not pushing directly to it anymore.  The non user-facing changes in here are:

* Use a read-only data class (like a namedtuple) to store cleaned user-provided data that was previously contained in attributes on the Stub classes.
* The `LocationTypeStub` and `LocationStub` classes are now initialized a little later in the process, and require a data object and the old collection as arguments to the constructor.  This allows for all methods and attributes to be effectively static.  Previously, they were initialized as part of the parsing process, and later injected with the old collection, which made for a state model that I found difficult to reason with.
* There was a test class that didn't depend on a db, but this required a substantial bit of mocking to support.  I made this a db test class and dropped the mocks.
* For another test class, I standardized the setup a bit by switching to `LocationHierarchyPerTest` rather than the one-off setup methods (I don't think that had existed at the time this was written).
* To support the change to require location_ids for updates, I put all test upload rows in either `NewLocRow`, which doesn't supply an ID, or `self.UpdateLocRow`, which looks up the ID by the site code.
* The location upload classes now require a user, which is used to make this feature location-safe.  I haven't yet marked it as such, since I want to do some UI work first.

@mkangia @millerdev @sravfeyn 

QA request: https://manage.dimagi.com/default.asp?279327

Oh also, I'm not crazy about the term "stub", let me know if you've got a better idea for what to call those things.

esoergel committed:
Get rid of all those nasty positional args

esoergel committed:
Remove a layer of indirection

esoergel committed:
Make user available to the uploader

esoergel committed:
Remove repetitive test setup

Many of these test methods used the same setup, so I split these up into
two classes - one with setup and one without.  I also converted the one
with setup to use the more standard LocationHierarchyPerTest class
rather than its reimplementation of it.

esoergel committed:
Split LocationTypeStub up

Now there's a namedtuple to represent the raw data and a helper class
for computing the upload.  The latter no longer mutates.

esoergel committed:
Move single-use data into the places it's referenced

esoergel committed:
Require location_id be provided for updates

esoergel committed:
use @attr.s for LocationTypeData

esoergel committed:
Iterate over old ancestor locations when deleting

esoergel committed:
If parent obj isn't being updated, use the old one

This sets parent at the very last minute, which will let us simplify
some of the other code considerably.  It also allows for the removal of
the fake_stub class, but I'm gonna commit that separately to avoid
conflicts.

esoergel committed:
Split LocationStub into a data class and helper class

This also takes the non-db tests and makes them db tests - there was a bunch of
mocking necessary to make them run, and it felt counter-productive.
I couldn't reasonably split this into multiple commits with the tests still
passing, sorry for the massive commit!

esoergel committed:
Allow for restricted location edit

esoergel committed:
remove `delay_updates`

esoergel committed:
use named args rather than positional

esoergel committed:
Parse to boolean in data model

esoergel committed:
Check site_codes against locs not in the upload

esoergel committed:
Clarify nomenclature a bit

esoergel committed:
Make tests provide loc IDs to update

This was failing after the new ID validation

esoergel committed:
lint

esoergel commented on corehq/apps/locations/tests/test_bulk_management.py:
```diff
@@ -609,21 +573,284 @@ def test_data_format(self):
         )
         assert_errors(result, ['index 0 should be valid decimal numbers'])
 
+    def test_custom_data(self):
```
This test and those below it are just being moved from that class into this one

esoergel commented on corehq/apps/locations/bulk_management.py:
```diff
@@ -878,6 +863,20 @@ def iter_unprocessed_ancestor_ids(stubs):
+                    location = old_locations_by_pk[pk]
+                    yield location.location_id
+                    pk = location.parent_id
+
```
Made a comment on the commit, but it didn't show up in diff view, so I'll copy here:

This change is twofold:

It's difficult to assign a parent location at the time when the location stub is being created. You'd need to either reorder locations to be top-to-bottom first (which is hard, since they're not yet fully parsed), or do two passes, one to init everything, then a second to add parents. I didn't like the second option since you'd have to mutate the location after its construction. Luckily, it looked like very little actually requires the parent object - pretty much just at the point of saving, so I opted to move the lookup there and drop the linkage.

The second aspect of this is that I realized we should be looking for the old ancestors here anyways, since they're the ones that are affected by a deletion.

mkangia replied:
Thanks Ethan... Can you add couple of comments within `save_locations` to understand whats happening here better.

millerdev reviewed (commented):
This is great! Left a few comments.

esoergel committed:
these attributes should always exist

esoergel committed:
PR feedback

esoergel committed:
Restrict creation of top-level locations

esoergel committed:
Don't allow passing in location ID to UpdateLocRow

I can't wait for Python3 keyword-only arguments

esoergel reviewed (commented):
just pushed some updates

millerdev reviewed (approved):
Excellent!

esoergel committed:
Give us a little more control over rendering the download section

esoergel committed:
Allow users to download blank, template upload

esoergel committed:
Prohibit restricted users from editing location types

esoergel committed:
'' means False

esoergel committed:
Order location types by hierarchy

esoergel committed:
Add test to upload the download

esoergel committed:
Add failing test and fix

esoergel committed:
lat/long should be decimals, coerce up-front

`needs_save` was incorrectly returning `True` because
`Decimal('42.1') != '42.1'`

esoergel committed:
Don't save a loc just to add empty metadata fields

esoergel committed:
Merge pull request #21129 from dimagi/es/location-upload2

Add empty template for uploads

esoergel committed:
Make location upload available to restricted users

mkangia commented on corehq/apps/locations/bulk_management.py:
```diff
@@ -403,14 +382,19 @@ def validate_and_parse_stubs_from_excel(self):
+    @staticmethod
+    def _get_type_data(index, row):
+        titles = LOCATION_TYPE_SHEET_HEADERS
+        name = row.get(titles['name'])
```
nit: `row.get(titles['name'], '').strip()` would help ensure empty spaces are removed.

esoergel replied:
Huh, it looks like this could use a fair bit more cleaning - I don't see anything that makes sure `site_codes` are valid slugs, either.  Thanks for flagging.

mkangia commented on corehq/apps/locations/bulk_management.py:
```diff
@@ -860,7 +861,14 @@ def append_at_bottom(parent_type):
-                    loc_object.parent = loc.new_parent
+                    parent_code = loc.parent_code
+                    if parent_code == ROOT_LOCATION_TYPE:
+                        loc_object.parent = None
```
Should not this raise an error instead of just removing the parent?

esoergel replied:
This isn't an error - the user is specifying that the location has no parent.

mkangia replied:
Aha.. so we assign the `parent`, the user just assigns the parent_code. Cool!

mkangia commented on corehq/apps/locations/tests/util.py:
```diff
@@ -158,16 +171,7 @@ def tearDownClass(cls):
-        role.save()
-        user.set_role(cls.domain, role.get_qualified_id())
-        user.save()
+        restrict_user_by_location(cls.domain, user)
```
`restrict_user_to_assigned_locations`/`remove_all_locations_access` sounds a better name considering we are not restricting user to any particular location but revoking `access_all_locations`?

mkangia commented on corehq/apps/locations/bulk_management.py:
```diff
@@ -568,6 +567,19 @@ def _check_unique_location_ids(self):
+              "All site_codes must be unique").format(code=l.site_code)
+            for l in self.all_listed_locations
+            if l.site_code in unavailable_site_codes
+        ]
```
this would have `All site_codes must be unique` in message for each unavailable site code? is that okay?

esoergel replied:
I think so, because the user needs to know which site codes are problematic.

mkangia replied:
I meant that in case there are two site codes that aren't unique, the messages would be
```
["Location site_code code1 is in use by another location. All site_codes must be unique,
   Location site_code code2 is in use by another location. All site_codes must be unique"]
```

esoergel replied:
That's consistent with the other error messages - they typically repeat the same message for each instance of the error.  I suppose the alternative would be something like "The following site_codes are are in use by another location: code1, code2".  I don't know, not sure I wanna go down that path in this PR.

mkangia commented:
build failure didnt look related. the test failing was `corehq.apps.api.tests:TestSimpleReportConfigurationResource.test_disallowed_methods`. Restated two jobs

millerdev commented on corehq/apps/locations/tests/test_bulk_management.py:
```diff
@@ -8,35 +8,33 @@
-from corehq.apps.locations.bulk_management import (
+from corehq.apps.users.models import WebUser
+
+from ..bulk_management import (
```
> Is it a good practice to used related modules with `...`?

@mkangia yes, if the code is expected to live at relatively the same location.

> If this file gets moved to a different location all of these would start failing right?

 The flip side of this question is if the `corehq.apps.locations` package gets moved or renamed then all non-relative internal imports will have to be updated.

I think it's good to use relative imports within packages where the files are all related in some clear way (for example, inside the `locations` app). However, not a good idea to do relative imports that span across apps or other major subdivisions of the codebase.

mkangia replied:
Cool, thanks @millerdev

esoergel commented on corehq/apps/locations/tests/test_bulk_management.py:
```diff
@@ -8,35 +8,33 @@
 
 FLAT_LOCATION_TYPES = [
-    # name, code, parent_code, do_delete, shares_cases, view_descendants, index
-    # ('name', 'code', 'parent_code', 'shares_cases', 'view_descendants'),
```
Because `LocationTypeData` is a class with named arguments now, so I didn't think it was as relevant.

mkangia replied:
okay. But one would need to first find that all these go as arguments to `LocationTypeData`.  Would be nice to still retain it to avoid that find.

esoergel commented on corehq/apps/locations/bulk_management.py:
```diff
@@ -787,14 +781,15 @@ def _check_model_validation(self):
     try:
-        importer = NewLocationImporter.from_excel_importer(domain, excel_importer)
+        validator = LocationExcelValidator(domain, excel_importer)
+        type_rows, location_rows = validator.validate_and_parse_stubs_from_excel()
```
These felt like two non-trivial phases of the process - one to parse the excel sheet into python data, and another one to actually run the import.  `NewLocationImporter` doesn't need to know anything about excel or the `LocationExcelValidator` class, and I thought this seemed like a more natural division.

esoergel committed:
use clearer syntax

esoergel committed:
Merge pull request #21136 from dimagi/es/location-download

Further location download/upload improvements

esoergel committed:
Document saving a little more

esoergel commented:
@millerdev @mkangia FYI I'm treating this branch as already approved, but I'm waiting on QA before merging.  Any updates are being PR'd into this branch, so it should be sufficient to review those PRs without re-reviewing this one.

mkangia commented on corehq/apps/locations/tests/test_bulk_management.py:
```diff
@@ -510,7 +461,7 @@ def bulk_update_locations(self, types, locations):
         # Makes sure that the set of all location types in the domain matches
         # the passed-in location types
-        actual_types = self.domain.location_types
+        actual_types = self.domain_obj.location_types
```
i see `domain_obj` being used already. Cool.

snopoke commented on corehq/apps/locations/views.py:
```diff
@@ -846,7 +846,10 @@ def post(self, request, *args, **kwargs):
         # the task will try to acquire.
-        task = import_locations_async.apply_async(args=[domain, file_ref.download_id], countdown=10)
+        task = import_locations_async.apply_async(
+            args=[domain, file_ref.download_id, request.couch_user],
```
It would be better to just pass the User ID and look it up in Celery rather than pass the entire user object.

snopoke commented:
Got part way through this. Will try review the rest later

esoergel committed:
Merge pull request #21157 from dimagi/es/location-upload-docs

Document saving a little more

esoergel committed:
Invalidate data definition class

This was causing an intermittent test failure.

A regrettable side effect of doing this was that I had to drop the import
statements from custom_data_fields.__init__.  This appears to be because of this
change from Django 1.9:

> All models need to be defined inside an installed application or declare an
explicit app_label. Furthermore, it isn’t possible to import them before their
application is loaded. In particular, it isn’t possible to import models inside
the root package of an application.

esoergel committed:
Pass only user_id rather than user object

esoergel committed:
strip string values in upload

esoergel committed:
Make light test wrapper around LocationTypeData

esoergel committed:
Why do you hate functional programming, Python?

The whole string module is deprecated - you're supposed to use methods now :(

esoergel committed:
Merge pull request #21172 from dimagi/es/location-upload-3

PR feedback from #21092

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into es/location-upload

esoergel commented on corehq/apps/locations/views.py:
```diff
@@ -865,6 +867,13 @@ def _cache_file(request, domain, upload):
+        # We need to start this task after this current request finishes because this
+        # request uses the lock_locations decorator which acquires the same lock that
+        # the task will try to acquire.
+        task = import_locations_async.apply_async(
```
this was a bad merge - now resolved

esoergel committed:
Test integer external IDs

excel will convert these columns to numeric types if they happen to be composed
of numerals

sravfeyn commented on corehq/apps/locations/bulk_management.py:
```diff
@@ -41,263 +47,194 @@ def success(self):
+
+
+@attrs(frozen=True)
+class LocationData(object):
```
Nice change!

esoergel committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into es/location-upload

esoergel committed:
Clear custom data cache on domain delete

esoergel commented:
Just implemented a proper fix for the caching issue that should fix the build: https://github.com/dimagi/commcare-hq/pull/21092/commits/9e42d494d0a8c9b60995733ba7273b17bd23ba83

orangejenny committed:
Merge branch 'master' into es/location-upload

esoergel committed:
lint [ci skip]
